DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/matmul/README.md

```diff
@@ -0,0 +1,323 @@
+
+<p align="center">
+  <img src=assets/overview.png />
+</p>
+
+Each iteration, a PE consumes inputs from previous PEs in the same row or column and uses these data to compute a multiply-accumulate. It then passes the inputs along to the next PE. We orchestrate the flow of data through the systolic array such that each PE is responsible for computing one element of the output matrix. Using this approach, it takes a total of *n* iterations to compute one matrix product, where every iteration, we compute a partial sum for each of the elements of the output matrix.
+
+<p align="center">
+  <img src=assets/systolic_array.png />
+</p>
+
+Here, a single PE consists of logic to compute a floating-point multiply-accumulate and registers for storage of intermediate results. 
+
+<p align="center">
+  <img src=assets/PE.png />
+</p>
+
+In our FPGA implementation, the multiply-accumulate operations are computed in a nested loop over the PEs. The loop is fully unrolled to replicate the hardware for each PE. As a result, *m<sub>PE</sub> × p<sub>PE</sub>* multiply-accumulate operations are performed in parallel on the FPGA. The multiply-accumulate arithmetic can be optimally implemented using the FPGA's specialized floating point DSP (Digital Signal Processing) hardware. When set to FP accum mode, we can perform this operation with an II of 1. With this optimization, our FPGA implementation requires *m<sub>PE</sub> × p<sub>PE</sub>* DSPs to compute all the floating-point multiply-accumulate operations. Thus, the matrix tile size is constrained by the total FPGA DSP resources available.
+
+To optimize the performance-critical loop in its algorithm, the design leverages concepts discussed in the following FPGA tutorials:
+
+- **Explicit Pipelining with `fpga_reg`** (fpga_register)
+- **Loop `ivdep` Attribute** (loop_ivdep)
+- **Unrolling Loops** (loop_unroll)
+
+The key optimization techniques used are as follows:
+
+1. Fully unrolling the loop over the matrix product to instantiate a systolic array of PEs
+2. Using an efficient memory banking scheme to generate high performance hardware.
+3. Using the `fpga_reg` attribute to insert more pipeline stages where needed to improve the frequency achieved by the design. This allows us to specify registers to pass inputs from one PE to the next. 
+
+### Compiler Flags Used
+
+| Flag                  | Description
+|:---                   |:---
+| `-Xshardware`         | Target FPGA hardware (as opposed to FPGA emulator)
+| `-Xsclock=360MHz`     | The FPGA backend attempts to achieve 360 MHz
```

Comment:
Without the clock target it seems the fmax is slightly lower, but I think the difference between the naive and systolic implementations is still roughly the same. Also I think this would just affect the A10 since I believe that is the only device where we set a different target than the default